pkg/nsx/services/ippool/ippool.go

```diff
@@ -0,0 +1,262 @@
+	tags := []model.Tag{
+		{Scope: String(common.TagScopeIPSubnetOwner), Tag: String(ResourceTypeIPPool)},
+	}
+	go ipPoolService.InitializeResourceStore(&wg, fatalErrors, ResourceTypeIPPoolBlockSubnet, tags, ipPoolService.ipPoolBlockSubnetStore)
+
+	go func() {
+		wg.Wait()
+		close(wgDone)
+	}()
+	select {
+	case <-wgDone:
+		break
+	case err := <-fatalErrors:
+		close(fatalErrors)
+		return ipPoolService, err
+	}
+	return ipPoolService, nil
+}
+
+func (service *IPPoolService) CreateOrUpdateIPPool(obj *v1alpha2.IPPool) (bool, bool, error) {
+	nsxIPPool, nsxIPSubnets := service.BuildIPPool(obj)
+	existingIPPool, existingIPSubnets, err := service.indexedIPPoolAndIPPoolSubnets(obj.UID)
+	if err != nil {
+		log.Error(err, "failed to get ip pool and ip pool subnets by UID", "UID", obj.UID)
+		return false, false, err
+	}
+	log.V(1).Info("existing ippool and ip subnets", "existingIPPool", existingIPPool, "existingIPSubnets", existingIPSubnets)
+	ipPoolSubnetsUpdated := false
+	ipPoolUpdated := common.CompareResource(IpAddressPoolToComparable(existingIPPool), IpAddressPoolToComparable(nsxIPPool))
+	changed, stale := common.CompareResources(IpAddressPoolBlockSubnetsToComparable(existingIPSubnets), IpAddressPoolBlockSubnetsToComparable(nsxIPSubnets))
+	changedIPSubnets, staleIPSubnets := ComparableToIpAddressPoolBlockSubnets(changed), ComparableToIpAddressPoolBlockSubnets(stale)
+	for i := len(staleIPSubnets) - 1; i >= 0; i-- {
+		staleIPSubnets[i].MarkedForDelete = &MarkedForDelete
+	}
+	finalIPSubnets := append(changedIPSubnets, staleIPSubnets...)
+	if len(finalIPSubnets) > 0 {
+		ipPoolSubnetsUpdated = true
+	}
+
+	if err := service.Operate(nsxIPPool, finalIPSubnets, ipPoolUpdated, ipPoolSubnetsUpdated); err != nil {
+		return false, false, err
+	}
+
+	realizedSubnets, subnetCidrUpdated, e := service.AcquireRealizedSubnetIP(obj)
+	if e != nil {
+		return false, false, e
+	}
+	obj.Status.Subnets = realizedSubnets
+	return subnetCidrUpdated, ipPoolSubnetsUpdated, nil
+}
+
+func (service *IPPoolService) Operate(nsxIPPool *model.IpAddressPool, nsxIPSubnets []*model.IpAddressPoolBlockSubnet, IPPoolUpdated bool, IPPoolSubnetsUpdated bool) error {
+	if !(IPPoolUpdated || IPPoolSubnetsUpdated) {
+		return nil
+	}
+	infraIPPool, err := service.WrapHierarchyIPPool(nsxIPPool, nsxIPSubnets)
+	if err != nil {
+		return err
+	}
+	// Get IPPool Type from nsxIPPool
+	IPPoolType := common.IPPoolTypePublic
+	for _, tag := range nsxIPPool.Tags {
+		if *tag.Scope == common.TagScopeIPPoolCRType {
+			IPPoolType = *tag.Tag
+			break
+		}
+	}
+
+	if IPPoolType == common.IPPoolTypePrivate {
+		ns := service.GetIPPoolNamespace(nsxIPPool)
+		VPCInfo := commonctl.ServiceMediator.ListVPCInfo(ns)
+		if len(VPCInfo) == 0 {
+			err = util.NoEffectiveOption{Desc: "no effective org and project for ippool"}
+		} else {
+			err = service.NSXClient.ProjectInfraClient.Patch(VPCInfo[0].OrgID, VPCInfo[0].ProjectID, *infraIPPool,
```

Comment:
Can we use IpPoolsClient under project infra? I see other PRs use  Resource clients, for example StaticRoutesClient.

Reply:
ProjectInfraClient use hierarchy mode